### ISSUE TITLE
关于attacks/gradient_method.py的问题

### ISSUE THREAD

**Issue by ygfrancois** [advboxes/AdvBox#24]
for i in range(steps):
                if adversary.is_targeted_attack:
                    gradient = +self.model.gradient(adv_img,
                                                    adversary.target_label)
                else:
                    gradient = -self.model.gradient(adv_img,
                                                   adversary.original_label)
                if norm_ord == np.inf:
                    gradient_norm = np.sign(gradient)
                else:
                    gradient_norm = gradient / self._norm(
                        gradient, ord=norm_ord)

                #logging.info('epsilon * gradient_norm={0}'.format(gradient_norm * epsilon))
                #logging.info('epsilon * gradient_norm* (max_ - min_)={0}'.format(gradient_norm * epsilon* (max_ - min_)))
                #改进的实现 不用考虑特征取值范围
                #adv_img = adv_img + epsilon * gradient_norm * (max_ - min_)
                #按照论文实现
                adv_img = adv_img + epsilon * gradient_norm
请问以上这段代码里.为什么**non_targeted_attack的时候加负号, targeted_attack的时候不加**呢?
我对算法的理解是non_targeted_attack情况下,应该梯度上升,由于后面是直接 + epsilon * gradient_norm,所以不明白这里为何要加个负号,这样不是反而变成梯度下降了么?

请教一下我哪里理解有问题,多谢了!

**Comment by wangaxe** [comment]
不知道对不对，希望能帮到你
```python
# pytorch版的model
def gradient(self, data, label):
         ...
         ...
        #loss=self._loss(output, label)
        ce = nn.CrossEntropyLoss()
        loss=-ce(output, label) #这里多了一个负号
```

**Comment by ygfrancois** [comment]
> 不知道对不对，希望能帮到你
> 
> ```python
> # pytorch版的model
> def gradient(self, data, label):
>          ...
>          ...
>         #loss=self._loss(output, label)
>         ce = nn.CrossEntropyLoss()
>         loss=-ce(output, label) #这里多了一个负号
> ```

谢谢回复，以下疑问：
 nn.CrossEntropyLoss()里应该包含了negative log likelihood里的负号了对吧？那这里这个loss，加一个负号的意义在哪呢？难道仅仅为了不同子类调用时候的一致么？如果是这样，意味着其他model例如tf的model也多加了一个负号？不过我没有找到其他框架的model在哪里加了负号，能否指示一下，多谢了。

另贴上 [cleverhans项目fast_gradient_method里target loss和非target loss的代码](https://github.com/tensorflow/cleverhans/blob/master/cleverhans/attacks/fast_gradient_method.py)

```
  # Compute loss
  loss = softmax_cross_entropy_with_logits(labels=y, logits=logits)
  if targeted:
    loss = -loss

  # Define gradient of loss wrt input
  grad, = tf.gradients(loss, x)

  optimal_perturbation = optimize_linear(grad, eps, ord)

  # Add perturbation to original example to obtain adversarial example
  adv_x = x + optimal_perturbation
```

**Comment by duoergun0729** [comment]
就是pytorch这个特殊点  其他的框架木有这个问题

**Comment by txyugood** [comment]
我今天也遇到这个问题了，我觉得这块应该是bug，我看 log 里面以前is_targeted_attack是减号，但是在ef59511b59a03755c33e712043f9c2211829191b 这个 commit 里面改了，所以能否查一下这是是否有问题？
